File: 646-maximum-length-of-pair-chain/646-maximum-length-of-pair-chain.cpp
class Solution {
public:
    int findLongestChain(vector<vector<int>>& pairs) {
        sort(pairs.begin(), pairs.end(), [](const vector<int>& p1, const vector<int>& p2) {
            return p1[0] < p2[0];
        });
        
        int res = 0;
        int n = pairs.size();
        vector<int> dp(n, 1);
        
        for (int i = 1; i < n; ++i) {
            for (int j = 0; j < i; ++j) {
                if (pairs[i][0] > pairs[j][1]) {
                    dp[i] = max(dp[i], dp[j] + 1);
                }
            }
        }
        
        for (int& i: dp) {
            res = max(res, i);
        }
        return res;
    }
};

File: 646-maximum-length-of-pair-chain/646-maximum-length-of-pair-chain.py
class Solution:
    def findLongestChain(self, pairs: List[List[int]]) -> int:
        pairs.sort(key=lambda x: x[0])
        
        def chainable(p1, p2):
            if p1[1] < p2[0]: return True
            return False
        
        n = len(pairs)
        dp = [1] * n
        
        for i in range(1, n):
            for j in range(i):
                if chainable(pairs[j], pairs[i]):
                    dp[i] = max(dp[i], dp[j] + 1)
        return max(dp)
        